File: this-is-coding-test/9.최단경로/2.미래_도시.py
node_count, main_count = 5, 7
data = [[1, 2], [1, 3], [1, 4], [2, 4], [3, 4], [3, 5], [4, 5]]
targets = [4, 5]
INFINITY = 987654321
graph = [[INFINITY] * (node_count+1) for _ in range(node_count+1)]


def show_graph(graph):
    graph = graph[1:]
    for index, row in enumerate(graph):
        print(row[1:])


for i in range(len(graph)):
    graph[i][i] = 0

show_graph(graph)

for [x, y] in data:
    graph[x][y] = 1
    graph[y][x] = 1

show_graph(graph)

for n in range(1, len(graph)):
    for x in range(1, len(graph)):
        for y in range(1, len(graph)):
            graph[x][y] = min(graph[x][y], graph[x][n]+graph[n][y])

show_graph(graph)

print(graph[1][targets[1]] + graph[targets[1]][targets[0]])

File: this-is-coding-test/9.최단경로/2.미래_도시.js
const node_count = 5;
const data = [
  [1, 2],
  [1, 3],
  [1, 4],
  [2, 4],
  [3, 4],
  [3, 5],
  [4, 5],
];
const targets = [5, 4];

const graph = Array.from({ length: node_count + 1 }, () =>
  Array.from({ length: node_count + 1 }, () => Infinity)
);

graph.forEach((_, i) => {
  graph[i][i] = 0;
  graph[i][0] = 0;
  graph[0][i] = 0;
});

console.log(1, graph);

data.forEach(([start, end]) => {
  graph[start][end] = 1;
  graph[end][start] = 1;
});

console.log(2, graph);

for (let n = 1; n <= node_count; n++) {
  for (let i = 1; i <= node_count; i++) {
    for (let j = 1; j <= node_count; j++) {
      graph[i][j] = Math.min(graph[i][j], graph[i][n] + graph[n][j]);
    }
  }
}

console.log(3, graph);

console.log(graph[1][5] + graph[5][4]);
